test/unit/autolinker_spec.js

```diff
@@ -0,0 +1,220 @@
+        "mailto:someone@subdomain.example.com",
+      ],
+      ["peter@abc.de", "mailto:peter@abc.de"],
+      ["red.teddy.b@abc.com", "mailto:red.teddy.b@abc.com"],
+      [
+        "abc_@gmail.com", // '_' is ok before '@'.
+        "mailto:abc_@gmail.com",
+      ],
+      [
+        "dummy-hi@gmail.com", // '-' is ok in user name.
+        "mailto:dummy-hi@gmail.com",
+      ],
+      [
+        "a..df@gmail.com", // Stop at consecutive '.'.
+        "mailto:a..df@gmail.com",
+      ],
+      [
+        ".john@yahoo.com", // Remove heading '.'.
+        "mailto:john@yahoo.com",
+      ],
+      [
+        "abc@xyz.org?/", // Trim ending invalid chars.
+        "mailto:abc@xyz.org",
+      ],
+      [
+        "fan{abc@xyz.org", // Trim beginning invalid chars.
+        "mailto:abc@xyz.org",
+      ],
+      [
+        "fan@g.com..", // Trim the ending periods.
+        "mailto:fan@g.com",
+      ],
+      [
+        "CAP.cap@Gmail.Com", // Keep the original case.
+        "mailto:CAP.cap@Gmail.Com",
+      ],
+    ];
+    const matches = Autolinker.findLinks(
+      emails.map(email => email[0]).join("\n")
+    );
+    expect(matches.length).toEqual(12);
+    expect(matches[0].url).toEqual(emails[0][1]);
+    expect(matches[1].url).toEqual(emails[1][1]);
+    expect(matches[2].url).toEqual(emails[2][1]);
+    expect(matches[3].url).toEqual(emails[3][1]);
+    expect(matches[4].url).toEqual(emails[4][1]);
+    expect(matches[5].url).toEqual(emails[5][1]);
+    expect(matches[6].url).toEqual(emails[6][1]);
+    expect(matches[7].url).toEqual(emails[7][1]);
+    expect(matches[8].url).toEqual(emails[8][1]);
+    expect(matches[9].url).toEqual(emails[9][1]);
+    expect(matches[10].url).toEqual(emails[10][1]);
+    expect(matches[11].url).toEqual(emails[11][1]);
+  });
+
+  it("should correctly handle complex or edge cases", function () {
+    const links = [
+      [
+        "https://example.com/path/to/page?query=param&another=val#section",
+        "https://example.com/path/to/page?query=param&another=val#section",
+      ],
+      [
+        "www.example.com/resource/(parentheses)-allowed/",
+        "http://www.example.com/resource/(parentheses)-allowed/",
+      ],
+      [
+        "http://example.com/path_with_underscores",
+        "http://example.com/path_with_underscores",
+      ],
+      [
+        "http://www.example.com:8080/port/test",
+        "http://www.example.com:8080/port/test",
+      ],
+      [
+        "https://example.com/encoded%20spaces%20in%20path",
+        "https://example.com/encoded%20spaces%20in%20path",
+      ],
+      ["mailto:hello+world@example.com", "mailto:hello+world@example.com"],
+      ["www.a.com/#a=@?q=rr&r=y", "http://www.a.com/#a=@?q=rr&r=y"],
+      ["http://a.com/1/2/3/4\\5\\6", "http://a.com/1/2/3/4/5/6"],
+      ["http://www.example.com/foo;bar", "http://www.example.com/foo;bar"],
+      // ["www.abc.com/#%%^&&*(", "http://www.abc.com/#%%^&&*("],
```

Comment:
Here, and elsewhere below, please remove such lines rather than leaving commented-out code around.

Reply:
They are edge cases from the PDFium test suite that we don't yet work with. They could be made to work at a later point.

Reply:
The test file has been streamlined but the commented out tests are still in there. I don't mind keeping them around personally since they are reminders about the edge cases we could work on but if you feel strongly about this I'll go back and remove them.

Reply:
You could just add something like `TODO: handle this case.` or something like that. An other possibility is to remove them as asked by @Snuffleupagus and once the patch is merged, file a bug with the different cases to be implemented.

@Snuffleupagus would it be fine for you ?

Reply:
As long as TODO comments are added, I suppose that fine since those can be found easily.

Reply:
This isn't addressed in the latest version.

Reply:
Addressed it.